crates/ty_server/src/server/api/traits.rs

```diff
@@ -1,22 +1,59 @@
-//! A stateful LSP implementation that calls into the ty API.
+//! Traits for handling requests and notifications from the LSP client.
+//!
+//! This module defines the trait abstractions used by the language server to handle incoming
+//! requests and notifications from clients. It provides a type-safe way to implement LSP handlers
+//! with different execution models (synchronous or asynchronous) and automatic retry capabilities.
+//!
+//! All request and notification handlers must implement the base traits [`RequestHandler`] and
+//! [`NotificationHandler`], respectively, which associate them with specific LSP request or
+//! notification types. These base traits are then extended by more specific traits that define
+//! the execution model of the handler.
+//!
+//! The [`SyncRequestHandler`] and [`SyncNotificationHandler`] traits are for handlers that
+//! executes synchronously on the main loop, providing mutable access to the [`Session`] that
+//! contains the current state of the server. This is useful for handlers that need to modify
+//! the server state such as when the content of a file changes.
+//!
+//! The [`BackgroundDocumentRequestHandler`] and [`BackgroundDocumentNotificationHandler`] traits
+//! are for handlers that operate on a single document and can be executed on a background thread.
+//! These handlers will have access to a snapshot of the document at the time of the request or
+//! notification, allowing them to perform operations without blocking the main loop. There is also
+//! the [`BackgroundRequestHandler`] trait for handlers that operate on the entire workspace
+//! instead of a single document and can also be executed on a background thread like fetching the
+//! workspace diagnostics.
```

Comment:
You may want to rename this section if you end up renaming `WorkspaceSnapshot` to `SessionSnapshot`.